File: solution/300. Longest Increasing Subsequence/300_02.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> q;
        for (int x : nums) {
            auto it = lower_bound(begin(q), end(q), x);
            if (it == q.end())
                q.push_back(x);
            else
                *it = x;
        }
        return q.size();
    }
};

File: solution/300. Longest Increasing Subsequence/300_02.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        A = []
        for x in nums:
            if not A or A[-1] < x:
                A.append(x)
            else:
                i = bisect.bisect_left(A, x);
                A[i] = x;
        return len(A)
    
